app/rest/controllers/src/main/java/io/syndesis/controllers/logs/LogsController.java

```diff
@@ -0,0 +1,379 @@
+                }
+
+                String name = pod.getMetadata().getName();
+                PodLogHandler handler = podHandlers.get(name);
+
+                if (handler == null) {
+                    // create a new handler.
+                    try {
+                        handler = new PodLogHandler(this, pod);
+                        handler.start();
+
+                        LOG.info("Created handler for pod: {}", handler.podName);
+                        podHandlers.put(name, handler);
+                    } catch (IOException e) {
+                        LOG.error("Unexpected Error", e);
+                    }
+                } else {
+                    // mark existing handlers as being used.
+                    handler.markInOpenshift.set(true);
+                }
+            }
+
+            // Remove items from the map which are no longer in openshift
+            Iterator<Map.Entry<String, PodLogHandler>> iterator = podHandlers.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<String, PodLogHandler> next = iterator.next();
+                if (!next.getValue().markInOpenshift.get()) {
+                    LOG.info("Pod not tracked by openshift anymore: {}", next.getValue().podName);
+                    next.getValue().keepTrying.set(false);
+                    iterator.remove();
+                }
+            }
+
+            @SuppressWarnings("unchecked")
+            HashMap<String, Object> pods = dbGet(HashMap.class, "/logs/pods");
+            if (pods != null) {
+                pods.keySet().removeAll(podHandlers.keySet());
+                for (String o : pods.keySet()) {
+                    jsonDB.delete("/logs/pods/" + o);
+                    LOG.info("Pod state removed from db: {}", o);
+                }
+            }
+        } catch (Throwable e) {
+            LOG.error("Unexpected Error occurred.", e);
+        }
+
+    }
+
+    protected PodList listPods() {
+        return client.pods().withLabel(OpenShiftService.COMPONENT_LABEL, "integration").list();
+    }
+
+    protected boolean isPodRunning(String name) {
+        Pod pod = client.pods().withName(name).get();
+        if (pod == null) {
+            return false;
+        }
+        return "Running".equals(pod.getStatus().getPhase());
+    }
+
+    /*
+     * Feeds the logs of the given podName to the callback handler for processing.
+     *
+     * We do this instead of using the watchLog() feature of the k8s client lib because it really sucks due to:
+     *  1. You can't configure the timestamps option or the sinceTime option.  Need to resume log downloads.
+     *  2. It seems to need extra threads..
+     *  3. It might be hiding some of the http failure conditions.
+     *
+     */
+    protected void getLogs(String podName, Callback handler, String sinceTime) throws IOException {
+        try {
+            PodOperationsImpl pod = (PodOperationsImpl) client.pods().withName(podName);
```

Comment:
This goes a long way, while it could just use something like:

    WatchLog w = client.pods().withName(podName).withTimestamps().watchLog();
    processLogStream(w.getOutput);

The part where you go down to the okhttp client, implement a custom callback etc (seems to me uneeded).

Reply:
See method comment:
```java
    /*
     * Feeds the logs of the given podName to the callback handler for processing.
     *
     * We do this instead of using the watchLog() feature of the k8s client lib because it really sucks due to:
     *  1. You can't configure the timestamps option or the sinceTime option.  Need to resume log downloads.
     *  2. It seems to need extra threads..
     *  3. It might be hiding some of the http failure conditions.
     *
     */
```

Has watchLog() resolved all of those?